### ISSUE TITLE
Make to_module preserve module state or require explicit opt-in for destructive tensor swaps

### ISSUE THREAD

**Issue by vmoens** [pytorch/tensordict#1719]
## Motivation

`TensorDict.to_module()` can currently be a footgun when the source TensorDict contains plain `Tensor` leaves and the destination module currently has `nn.Parameter` or buffer registrations.

In that situation, applying the TensorDict may replace registered module state with plain tensors. A downstream symptom is that `module.state_dict()` can lose entries after a weight sync, even though the module may still run.

This surfaced in TorchRL evaluator / collector weight synchronization where weights are often represented as detached TensorDicts, for example from `TensorDict.from_module(policy).data.detach().clone().cpu()`. Reapplying those weights should be idempotent and should not destroy the receiving module's `state_dict()`.

## Desired behavior

We should tighten the `to_module()` semantics around destructive replacements:

1. Warn or error by default when `to_module()` would replace an existing `nn.Parameter` or registered buffer with a plain tensor, or otherwise perform an operation that would remove entries from `state_dict()`.
2. Require an explicit opt-in kwarg for the current destructive behavior, e.g. `allow_plain_tensor_module_state=True` / `allow_state_dict_mutation=True` / similar.
3. Provide a safe mode that auto-casts source leaves according to the destination module state:
   - destination `nn.Parameter` stays an `nn.Parameter` with `requires_grad` preserved;
   - destination buffer stays a registered buffer;
   - plain attributes can remain plain tensors.
4. Keep and document an in-place copy mode for weight syncs. I believe `inplace=True` already copies into existing storage in many cases, but we should make this behavior explicit and test it as the recommended idempotent path.

## API direction

Possible approaches:

- Add a new explicit API such as `TensorDict.copy_to_module_()` / `TensorDict.update_module_()` for idempotent weight syncs.
- Or add kwargs to `to_module()`, for example:
  - `preserve_module_state=True` to cast by destination registration and preserve `state_dict()` keys;
  - `allow_plain_tensor_replacement=False` by default, requiring users to opt into the old replacement behavior.

The main requirement is that users have a safe, obvious way to apply TensorDict weights to a module without accidentally deregistering parameters or buffers, and that unsafe behavior is not silent.

## Minimal pattern

```python
import torch.nn as nn
from tensordict import TensorDict

module = nn.Linear(4, 2)
weights = TensorDict.from_module(module).data.detach().clone()

# This kind of plain-tensor weight sync should not silently destroy
# module parameter/buffer registration or make state_dict() lose keys.
weights.to_module(module)

assert module.state_dict()
```

## Related downstream issue

This behavior caused a TorchRL evaluator weight-sync bug: https://github.com/pytorch/rl/issues/3804
